File: Python/roman-to-integer.py
class Solution:
    # @param {string} s
    # @return {integer}
    def romanToInt(self, s):
        ans = 0
        pre_c2n = 100000
        dic = {'I':1, 'V':5, 'X':10, 'L':50, 'C':100, 'D':500, 'M':1000}
        
        for c in s:
            cur_c2n = dic[c]
            ans += cur_c2n
            
            if pre_c2n < cur_c2n:
                ans -= pre_c2n<<1
            
            pre_c2n = cur_c2n
            
        return ans

File: C++/roman-to-integer.cpp
class Solution {
public:
    int romanToInt(std::string s) {
        std::unordered_map<char, int> dic({{'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}});
        int ans = 0;
        int prev_c2n = 10000;
        int curr_c2n = 0;
        for(char c : s) {
            curr_c2n = dic[c];
            ans += curr_c2n;
            if(prev_c2n<curr_c2n) {
                ans -= prev_c2n<<1;
            }
            prev_c2n = curr_c2n;
        }
        return ans;
    }
};

